File: mobile/src/screens/AlarmsScreen.js
import React, { useState, useEffect } from 'react';
import {
  View,
  Text,
  FlatList,
  StyleSheet,
  RefreshControl,
  ActivityIndicator,
  TouchableOpacity,
} from 'react-native';
import { fetchLatestAlarms } from '../services/apiService';
import { addAlarmListener } from '../services/socketService';

export default function AlarmsScreen() {
  const [alarms, setAlarms] = useState([]);
  const [isLoading, setIsLoading] = useState(true);
  const [isRefreshing, setIsRefreshing] = useState(false);
  const [filter, setFilter] = useState('all');

  useEffect(() => {
    loadAlarms();

    // Listen for real-time updates
    const removeListener = addAlarmListener((newAlarm) => {
      setAlarms(prev => [newAlarm, ...prev]);
    });

    return () => removeListener();
  }, []);

  const loadAlarms = async () => {
    try {
      const data = await fetchLatestAlarms(100);
      setAlarms(data);
    } catch (error) {
      console.error('Error loading alarms:', error);
    } finally {
      setIsLoading(false);
      setIsRefreshing(false);
    }
  };

  const onRefresh = () => {
    setIsRefreshing(true);
    loadAlarms();
  };

  const getFilteredAlarms = () => {
    if (filter === 'all') return alarms;
    return alarms.filter(alarm => alarm.source === filter);
  };

  const renderAlarmItem = ({ item }) => (
    <View style={styles.alarmCard}>
      <View style={styles.alarmHeader}>
        <View style={[styles.sourceBadge, { backgroundColor: getSourceColor(item.source) }]}>
          <Text style={styles.sourceBadgeText}>{formatSourceName(item.source)}</Text>
        </View>
        <Text style={styles.alarmTime}>
          {formatDateTime(item.received_at)}
        </Text>
      </View>
      <Text style={styles.alarmMessage}>{item.message}</Text>
      {item.raw_data && item.raw_data !== item.message && (
        <Text style={styles.rawData}>Raw: {item.raw_data}</Text>
      )}
    </View>
  );

  if (isLoading) {
    return (
      <View style={styles.centered}>
        <ActivityIndicator size="large" color="#007AFF" />
      </View>
    );
  }

  return (
    <View style={styles.container}>
      <View style={styles.filterContainer}>
        <TouchableOpacity
          style={[styles.filterButton, filter === 'all' && styles.filterButtonActive]}
          onPress={() => setFilter('all')}
        >
          <Text style={[styles.filterText, filter === 'all' && styles.filterTextActive]}>
            All
          </Text>
        </TouchableOpacity>
        <TouchableOpacity
          style={[styles.filterButton, filter === 'serial' && styles.filterButtonActive]}
          onPress={() => setFilter('serial')}
        >
          <Text style={[styles.filterText, filter === 'serial' && styles.filterTextActive]}>
            Serial
          </Text>
        </TouchableOpacity>
        <TouchableOpacity
          style={[styles.filterButton, filter === 'tap' && styles.filterButtonActive]}
          onPress={() => setFilter('tap')}
        >
          <Text style={[styles.filterText, filter === 'tap' && styles.filterTextActive]}>
            TAP
          </Text>
        </TouchableOpacity>
        <TouchableOpacity
          style={[styles.filterButton, filter === 'serial_ip' && styles.filterButtonActive]}
          onPress={() => setFilter('serial_ip')}
        >
          <Text style={[styles.filterText, filter === 'serial_ip' && styles.filterTextActive]}>
            Serial/IP
          </Text>
        </TouchableOpacity>
      </View>

      <FlatList
        data={getFilteredAlarms()}
        renderItem={renderAlarmItem}
        keyExtractor={(item) => item.id.toString()}
        contentContainerStyle={styles.listContent}
        refreshControl={
          <RefreshControl refreshing={isRefreshing} onRefresh={onRefresh} />
        }
        ListEmptyComponent={
          <Text style={styles.emptyText}>No alarms found</Text>
        }
      />
    </View>
  );
}

const getSourceColor = (source) => {
  const colors = {
    serial: '#ff6b6b',
    tap: '#4ecdc4',
    serial_ip: '#ffe66d',
  };
  return colors[source] || '#999';
};

const formatSourceName = (source) => {
  const names = {
    serial: 'Serial',
    tap: 'TAP',
    serial_ip: 'Serial/IP',
  };
  return names[source] || source;
};

const formatDateTime = (timestamp) => {
  const date = new Date(timestamp);
  return date.toLocaleString();
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#f5f5f5',
  },
  centered: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  filterContainer: {
    flexDirection: 'row',
    padding: 15,
    backgroundColor: '#fff',
    borderBottomWidth: 1,
    borderBottomColor: '#e0e0e0',
  },
  filterButton: {
    flex: 1,
    paddingVertical: 8,
    paddingHorizontal: 12,
    marginHorizontal: 4,
    borderRadius: 20,
    backgroundColor: '#f0f0f0',
    alignItems: 'center',
  },
  filterButtonActive: {
    backgroundColor: '#007AFF',
  },
  filterText: {
    fontSize: 14,
    fontWeight: '600',
    color: '#666',
  },
  filterTextActive: {
    color: '#fff',
  },
  listContent: {
    padding: 15,
  },
  alarmCard: {
    backgroundColor: '#fff',
    borderRadius: 12,
    padding: 15,
    marginBottom: 10,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 4,
    elevation: 3,
  },
  alarmHeader: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginBottom: 10,
  },
  sourceBadge: {
    paddingHorizontal: 10,
    paddingVertical: 4,
    borderRadius: 12,
  },
  sourceBadgeText: {
    color: '#fff',
    fontSize: 12,
    fontWeight: '600',
  },
  alarmTime: {
    fontSize: 12,
    color: '#999',
  },
  alarmMessage: {
    fontSize: 16,
    color: '#333',
    marginBottom: 5,
  },
  rawData: {
    fontSize: 12,
    color: '#999',
    fontStyle: 'italic',
  },
  emptyText: {
    textAlign: 'center',
    color: '#999',
    fontSize: 16,
    marginTop: 50,
  },
});

File: mobile/lib/screens/alarms_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:intl/intl.dart';
import '../services/app_state.dart';
import '../models/alarm.dart';

class AlarmsScreen extends StatefulWidget {
  const AlarmsScreen({super.key});

  @override
  State<AlarmsScreen> createState() => _AlarmsScreenState();
}

class _AlarmsScreenState extends State<AlarmsScreen> {
  String _selectedFilter = 'all';

  Color _getSourceColor(String source) {
    switch (source) {
      case 'serial':
        return const Color(0xFFFF6B6B);
      case 'tap':
        return const Color(0xFF4ECDC4);
      case 'serial_ip':
        return const Color(0xFFFFE66D);
      default:
        return Colors.grey;
    }
  }

  String _formatSourceName(String source) {
    switch (source) {
      case 'serial':
        return 'Serial';
      case 'tap':
        return 'TAP';
      case 'serial_ip':
        return 'Serial/IP';
      default:
        return source;
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Alarms'),
        backgroundColor: Colors.blue,
        foregroundColor: Colors.white,
      ),
      body: Consumer<AppState>(
        builder: (context, appState, child) {
          final alarms = appState.getFilteredAlarms(_selectedFilter);

          return Column(
            children: [
              // Filter Buttons
              Container(
                padding: const EdgeInsets.all(16),
                decoration: const BoxDecoration(
                  color: Colors.white,
                  border: Border(
                    bottom: BorderSide(color: Colors.grey, width: 0.5),
                  ),
                ),
                child: Row(
                  children: [
                    _FilterButton(
                      label: 'All',
                      isSelected: _selectedFilter == 'all',
                      onTap: () => setState(() => _selectedFilter = 'all'),
                    ),
                    const SizedBox(width: 8),
                    _FilterButton(
                      label: 'Serial',
                      isSelected: _selectedFilter == 'serial',
                      onTap: () => setState(() => _selectedFilter = 'serial'),
                    ),
                    const SizedBox(width: 8),
                    _FilterButton(
                      label: 'TAP',
                      isSelected: _selectedFilter == 'tap',
                      onTap: () => setState(() => _selectedFilter = 'tap'),
                    ),
                    const SizedBox(width: 8),
                    _FilterButton(
                      label: 'Serial/IP',
                      isSelected: _selectedFilter == 'serial_ip',
                      onTap: () => setState(() => _selectedFilter = 'serial_ip'),
                    ),
                  ],
                ),
              ),

              // Alarms List
              Expanded(
                child: RefreshIndicator(
                  onRefresh: () => appState.loadData(),
                  child: alarms.isEmpty
                      ? const Center(
                          child: Text(
                            'No alarms found',
                            style: TextStyle(
                              fontSize: 16,
                              color: Colors.grey,
                            ),
                          ),
                        )
                      : ListView.builder(
                          padding: const EdgeInsets.all(16),
                          itemCount: alarms.length,
                          itemBuilder: (context, index) {
                            final alarm = alarms[index];
                            return _AlarmCard(
                              alarm: alarm,
                              getSourceColor: _getSourceColor,
                              formatSourceName: _formatSourceName,
                            );
                          },
                        ),
                ),
              ),
            ],
          );
        },
      ),
    );
  }
}

class _FilterButton extends StatelessWidget {
  final String label;
  final bool isSelected;
  final VoidCallback onTap;

  const _FilterButton({
    required this.label,
    required this.isSelected,
    required this.onTap,
  });

  @override
  Widget build(BuildContext context) {
    return Expanded(
      child: GestureDetector(
        onTap: onTap,
        child: Container(
          padding: const EdgeInsets.symmetric(vertical: 8),
          decoration: BoxDecoration(
            color: isSelected ? Colors.blue : Colors.grey[200],
            borderRadius: BorderRadius.circular(20),
          ),
          child: Text(
            label,
            style: TextStyle(
              fontSize: 14,
              fontWeight: FontWeight.w600,
              color: isSelected ? Colors.white : Colors.grey[700],
            ),
            textAlign: TextAlign.center,
          ),
        ),
      ),
    );
  }
}

class _AlarmCard extends StatelessWidget {
  final Alarm alarm;
  final Color Function(String) getSourceColor;
  final String Function(String) formatSourceName;

  const _AlarmCard({
    required this.alarm,
    required this.getSourceColor,
    required this.formatSourceName,
  });

  @override
  Widget build(BuildContext context) {
    return Card(
      margin: const EdgeInsets.only(bottom: 10),
      elevation: 2,
      child: Padding(
        padding: const EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                Container(
                  padding: const EdgeInsets.symmetric(
                    horizontal: 10,
                    vertical: 4,
                  ),
                  decoration: BoxDecoration(
                    color: getSourceColor(alarm.source),
                    borderRadius: BorderRadius.circular(12),
                  ),
                  child: Text(
                    formatSourceName(alarm.source),
                    style: const TextStyle(
                      color: Colors.white,
                      fontSize: 12,
                      fontWeight: FontWeight.w600,
                    ),
                  ),
                ),
                Text(
                  DateFormat('yyyy-MM-dd HH:mm:ss').format(alarm.receivedAt),
                  style: const TextStyle(
                    fontSize: 12,
                    color: Colors.grey,
                  ),
                ),
              ],
            ),
            const SizedBox(height: 10),
            Text(
              alarm.message,
              style: const TextStyle(fontSize: 16),
            ),
            if (alarm.rawData != null && alarm.rawData != alarm.message) ...[
              const SizedBox(height: 8),
              Text(
                'Raw: ${alarm.rawData}',
                style: const TextStyle(
                  fontSize: 12,
                  color: Colors.grey,
                  fontStyle: FontStyle.italic,
                ),
              ),
            ],
          ],
        ),
      ),
    );
  }
}
